server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java

```diff
@@ -212,14 +215,25 @@ public void testForceStaleReplicaToBePromotedToPrimary() throws Exception {
                 rerouteBuilder.add(new AllocateEmptyPrimaryAllocationCommand(idxName, shardId, storeStatus.getNode().getId(), true));
             }
         }
-        rerouteBuilder.get();
-
-        ClusterState state = client().admin().cluster().prepareState().get().getState();
 
-        Set<String> expectedAllocationIds = useStaleReplica
+        final Set<String> expectedAllocationIds = useStaleReplica
             ? Collections.singleton(RecoverySource.ExistingStoreRecoverySource.FORCED_ALLOCATION_ID)
             : Collections.emptySet();
-        assertEquals(expectedAllocationIds, state.metaData().index(idxName).inSyncAllocationIds(0));
+
+        final CountDownLatch clusterStateChangeLatch = new CountDownLatch(1);
+        final ClusterStateListener clusterStateListener = event -> {
+            final Set<String> allocationIds = event.state().metaData().index(idxName).inSyncAllocationIds(0);
+            if (expectedAllocationIds.equals(allocationIds)) {
+                clusterStateChangeLatch.countDown();
+            }
+        };
+        final ClusterService clusterService = internalCluster().getInstance(ClusterService.class, master);
+        clusterService.addListener(clusterStateListener);
+
+        rerouteBuilder.get();
+
+        assertThat(clusterStateChangeLatch.await(30, TimeUnit.SECONDS), equalTo(true));
```

Comment:
Why not just `assertTrue()`?

Reply:
agreed: there are ~3250 `assertTrue`+`assertFalse` vs 1300 `equalTo(true)`/`equalTo(false)`

addressed in 22a6ffb